### PR TITLE
ENH: Add the __pow__ operator for Rotations

### LINKED ISSUE #17429: ENH: Add power (exponent) operator for Rotation objects
### Is your feature request related to a problem? Please describe.

At the moment, scipy.spatial.transform.Rotation objects can be composed with the multiplication operator. The power operator is a natural extension of this to compose a rotation with itself (ie, apply the same rotation multiple times).

### Describe the solution you'd like.

For an exponent `n` where `n` is a positive integer, the interpretation is straightforward:
`R * R == R**2`

For `n=1`, the original rotation should be returned.
`R == R**1`

For `n=0`, the identity rotation should be returned, but I'm open to arguments that this should be left undefined.
`Rotation.from_quat([0, 0, 0, 1]) == R**0`

For integer `n<0`, the obvious interpretation would be to return the inverse rotation, but I'm more ambivalent on whether this is implemented or left undefined. A future improvement could be to add both this and the division operator as a representation for inversion.
`R.inv() * R.inv() == R**-2`
`R1 / R2 = R1 * R2.inv()`

For non-integer `n`, we could consider it as a 'scaling' of the rotation, and apply `n` as a factor to the angle in the rotation vector interpretation. However, this results in underconstrained behavior. I think it makes more sense to leave it undefined for now.
For example, consider a 180 deg rotation raised to a 3/2 power, where it is unclear in which direction to apply the additional 90 deg: 
```python
R1 = Rotation.from_rotvec(np.pi * np.array([0, 0, 1]))
R2 = Rotation.from_rotvec(-np.pi * np.array([0, 0, 1]))
R1 == R2
R1a = R1 ** 3/2 == Rotation.from_rotvec(3/2 * np.pi * np.array([0, 0, 1]))
R2a = R2 ** 3/2 == Rotation.from_rotvec(-3/2 * np.pi * np.array([0, 0, 1]))
R1a != R2a
```

### Describe alternatives you've considered.

Instead of `R*R == R**2`, one might be tempted to write `R*R == 2*R`. However this introduces confusion since in many of the representations of a rotation multiplying all the factors by a scalar does not change the meaning, and this breaks with the current analogy of multiplication standing in for composition.

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by scottshambaugh**
#### Reference issue
Closes https://github.com/scipy/scipy/issues/17429

#### What does this implement/fix?
This adds the `__pow__` operator for scipy.spatial.transform.Rotation objects. This allows them to be composed with themselves, for example: `R * R == R ** 2`

Non-integer powers work the same as scaling the magnitude of the angle in the rotvec representation. This means that a power of 0 returns the identity rotation.

~~I have required the exponent *n* to be nonnegative.~~ If there's interest, then negative numbers could return the inverse rotation raised to a power. In this instance, `R ** -2 == R.inv() ** 2` Update: this has been updated.

#### Additional information
Cython requires `__pow__` to include a third `modulus` argument, which is not relevant for this use case. This ([stackoverflow question](https://stackoverflow.com/questions/56808420/cython-make-the-modulo-argument-optional-in-pow)) shows that it should default to None, which I add a check for, but I'm not sure how to gracefully handle this unused parameter in the docstring.

The taylor series expansion for small angles [was calculated with wolfram alpha](https://www.wolframalpha.com/input?key=&i=taylor+series+sin%28n+*+x+%2F+2%29+%2F+sin%28x+%2F+2%29).

### CODE REVIEW

**Comment by scottshambaugh** [comment]
This depends on https://github.com/scipy/scipy/pull/17334 being merged in first for tests to pass.

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [inline scipy/spatial/transform/_rotation.pyx:1946]: Maybe a silly question, but why not implement negative powers as the inverse rotation?
Since, for example `q**(-2) = (q**2)**(-1)`

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1946]: I simply hadn’t seen that notation anywhere before, but it makes sense to me and as you say should be easy to add.

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [inline scipy/spatial/transform/_rotation.pyx:1973]: Another silly question, did you check if there is an actual gain in execution speed? Apart from making the code slightly less readable, I'd imagine this kind of optimization is probably already inside of the cython implementations of `sin`...

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [inline scipy/spatial/transform/_rotation.pyx:1961]: Since this is not performed for the cases `n == 2` and `n == 1`, wouldn't this cause some minor inconsistent behavior?

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1973]: I didn’t check to see if premultiplying n3 and n5 saved any time, but the Taylor series expansion here is to protect for a divide by 0 error/instability rather than as an optimization.

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1961]: There will be some slightly inconsistent behavior for the n=2 case, for a 90 deg rotation that gets squared to the opposite pole. See some discussion [here](https://github.com/scipy/scipy/pull/17334) about that case (aka the q[3] == 0 case). I think it’s not so much an issue for n=2 since it’s a marginal epsilon scenario, as it might result in different representations for n=2 and n=5 (which for a 90 deg rotation should be the same).

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1973]: >  the Taylor series expansion here is to protect for a divide by 0 error/instability rather than as an optimization.

Ok I see it now! :)

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1961]: I see what you mean! 
Personally I'd prefer to do the same to all of them for consistency, specially since `n=2` could be possible the most used case.

File: scipy/spatial/transform/_rotation.pyx
Comment (rkern) [inline scipy/spatial/transform/_rotation.pyx:1983]: Should this be one indentation level up?

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1946]: This would imply to me that we should also implement the `__div__` operator, such that the following are equivalent: `A / B == A * B.inv()`.
@rkern have you see that notation anywhere?

Comment (rkern) [reply scipy/spatial/transform/_rotation.pyx:1946]: Less so because of non-commutativity. It's easier to express the right things with `*` and negative powers.

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1973]: Here's another idea for this: do we know about what value for `norm(quat[:3])` the zero division starts to create numerical problems?

An interesting problem could be to analyze how the error between `sin(n*angle/2) / sin(angle/2)` changes with the number of terms of the Taylor series, according to the distance to `0`, and then use a smaller `eps` so that we can simply state `sin(n*angle/2) / sin(angle/2) = n`

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [inline scipy/spatial/transform/_rotation.pyx:1969]: Here's another idea for this: 

Do we know about what value for `norm(quat[:3])` the zero division starts to create numerical problems?

An interesting problem could be to analyze how the error between `sin(n*angle/2) / sin(angle/2)` changes with the number of terms of the Taylor series, according to the distance to `0`, and then use a smaller `eps` (smaller than `1e-3`) so that we can simply state `sin(n*angle/2) / sin(angle/2) = n`

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1969]: I just did a quick test. Defined these functions:
```
def f(theta,n):
    return np.sin(n*theta/2) / np.sin(theta/2)

def g(theta,n):

    n3 = n*n*n
    n5 = n*n*n3
    return n + (n - n3) / 24 * theta + (3 * n5 - 10 * n3 + 7 * n) / 5760 * theta * theta
```
And then did some tests with different values:
```
f(1e-10, 100) == 100
g(1e-10, 100) == 99.99999583375006
f(1e-10, 1000) == 999.9999999999997
g(1e-10, 1000) == 999.9958333427084
```
We can see that the original function `f` gives values closer to the `n`. So I really think that lowering the `eps` value for the angle and then reducing the Taylor series to a simple `n` can actually give a more accurate result, apart from been a lil bit faster and a lot cleaner.

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1969]: Alternatively, if there is a Cython implementation of the `sinc` function (if there is, I'll suppose it's defined as the numpy version, which is `sinc(x) = sin(pi*x) / (pi*x)`), then for small `angle`:
```
sin(n * angle / 2) / sin(angle / 2) 
    == sin(n * angle / 2) / (angle / 2) 
    == n * sin(n * angle / 2) / (n * angle / 2)
```
Which gives:
```
sin(n * angle / 2) / sin(angle / 2) == n * sinc(n * angle / (2 * pi))
```
This would probably be better, since it would lower the error caused by large values of `n` when taking only the linear term of the sine Taylor expansion.

Comment (WarrenWeckesser) [reply scipy/spatial/transform/_rotation.pyx:1969]: FYI: The function $$f(x, n) = \frac{\sin(n x / 2)}{n \sin(x/2)}$$ is known as the periodic sinc function or the Dirichlet function.  We have an implementation at [`scipy.special.diric`](https://scipy.github.io/devdocs/reference/generated/scipy.special.diric.html).  I haven't checked if that provides any advantage over the code in this PR--the factor of `n`  in `n*diric(angle, n)` might multiply the numerical error in an undesirable way.  (Also note that `diric` requires `n > 0`, so an additional line or two of code would be needed to handle `n <= 0`.)

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1969]: Interesting, this is even better! 
In this PR's implementation, we can safely assume that `n > 0` by this stage.

For the numerical error, I think it would be better, but alternatively an extra case could be added:
```
        if angle <= 1e-7 and n <= 1e-3:
                    scale = n
        elif angle <= 1e-7:
                    scale = n * diric(angle, n)
        else:  # large angle
                    scale = sin(n * angle / 2) / sin(angle / 2)
```

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1969]: That's an interesting look @evbernardes, thanks for running that! There are two other taylor series expansions in this file, they would probably benefit from another PR with the same sort of look.

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1969]: Alright, just checked it. For the first one:
```
sin(angle / 2) / angle == sinc(angle / (2 * pi)) / 2
```
Or, using `diric` as pointed out by @WarrenWeckesser :
```
sin(angle / 2) / angle == diric(angle, 1) / 2
```
I think you could just replace everything by that instead of having the case checking.

Alternatively, since here we don't have the problem of having the `n` factor that can also vary, I think in this case you can just use a the smaller `eps` and and use the first term of the Taylor expansion for small angles: `sinc(angle / 2) / angle == 1/2`.

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1969]: For the other case, it's basically the reciprocal of that... 
So either remove case checking and:
```
angle / sin(angle / 2)  == 2 / sinc(angle / (2 * pi))
angle / sin(angle / 2)  == 2 / diric(angle, 1)
```
Or keep case checking, lower `eps` and for small values:
```
angle / sin(angle / 2)  == 2
```
I haven't seen the code much, but I imagine there are no other singularity problems with dividing by sine, since `angle/2` only grows up to 90 degrees, right?

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1961]: Just a last comment on this, I've quickly read through your PR about mapping the quaternion to a single cover.

I'd argue that the fact it hasn't been decided (at least for now) whether to merge it or not is a good argument to remove this from here, and let the decision be made elsewhere in the code, as a parameter, etc.

Even if the single cover mapping does get merged, but then someday in the future for some reason it gets reversed, it would probably be better not to have this kind of leftover behavior.

What do you think?

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1961]: The single cover mapping here is a quick way to take the shortest angle path from one point to another, rather than accidentally taking the opposite direction `2*pi - angle` path. The same logic is in `as_rotvec()`. It’s necessary for proper functioning, and unrelated to the single cover MR which is about how to represent things.

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1961]: It will be inconsistent in the double cover quaternion space, yes. It will be the same in every other representation. In general I don’t think we should rely on the properties of the double cover space since the abstract object here is a Rotation rather than a quaternion, and users might only be familiar with one of the representations (ie rotation matrices). If we weren’t to enforce the shortest path difference then we would have a well defined operation on the quaternions, but we would force users into using quaternions. This could result in unexpected behavior if mixing representations. I’m all ears though if there’s a clever way to mitigate this.

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1961]: I do think it would be a good idea to scope out some documentation and tests to define the explicit scope of where the double cover properties are retained. Likely just `__mul__`, `.inv()`, `.from_quat()`, and `.as_quat(canonical=False)`. Probably a separate PR though.

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1961]: I see what you mean, yes! A agree that the choice should be made focused on how we "see" it as an abstract "Rotation" and not as a quaternion. What I mean though is that if this gets accepted, maybe some thought should be given in whether it would be beneficial to also apply this to the simple quaternion composition.

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1961]: Maybe, I think it warrants some more discussion.

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1961]: Personally I would prefer to leave both things completely separated. 

I'd define __pow__ as the direct quaternion power, for consistency with __add__ being equivalent to a direct quaternion composition. Mostly, because I think this inconsistency might also create bugs that are hard to analyze for someone not familiar with how they are defined.

Then maybe in the class initialiser we could have a final check that always assures the stored quaternion has positive single part (except if a special flag is set to False, as you said!)

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1961]: I haven't taken out the `n == 2` case here yet but once I do I'm not sure I'm following what your recommended change is?

Comment (evbernardes) [reply scipy/spatial/transform/_rotation.pyx:1961]: Sorry if I'm mistaken, I'm currently seeing this from my phone. Here's what I mean:

Currently, the code first maps the quaternion to the canonical cover, then calculates the angle, then multiplies the angle by n.

This multiplied angle can be bigger than pi though, and no more mapping is made after that, so the resulting quaternion can be possibly on the non canonical cover. Am I right?

What I propose is to first remove the check before multiplying the angle, and then, let all the mapping happen at the very end, when calling `__class__`.

In this case, the class initialiser would deal with the cover mapping.

Again sorry if I'm mistaken, hopefully I managed to be clearer!

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1961]: I think I see what you mean! But I think this may run into issues with fractional powers. Let me play with it when I get home.

Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx]: Removed the `n==2` case and added support for negative exponents.

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx]: Added support for negative exponents.

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx]: I updated the approximation introduced in this PR, but I think changes to the other two should be broken out separately. I opened this issue to track it: https://github.com/scipy/scipy/issues/17635

**Comment by evbernardes** [comment]
For PR #17642, I tried using `sinc` instead of a case checking with the Taylor coefficients and it was much slower.

Perhaps using `diric` here will have the same problem, so I was wondering if is this implementation faster than simply implementing:

```python
def __pow__(Rotation self, float n, modulus):
    return Rotation.from_rotvec(n * self.as_rotvec())
```
Like you already implemented on the tests? If this is faster, then maybe this could be implemented instead, since it is much simpler.

**Comment by scottshambaugh** [comment]
I could use a hand debugging the docs build error from someone more familiar with Sphinx. Even when I build locally verbose I don't get more than this, and for the life of me I cannot figure out where the error is.

```
/home/circleci/repo/doc/docstring of scipy.spatial.transform._rotation.Rotation:233:<autosummary>:1: WARNING: py:obj reference target not found: scipy.spatial.transform.Rotation.__pow__
```

**Comment by evbernardes** [comment]
> @evbernardes I switch the logic over to what you suggested. I didn't check speed, but it certainly makes the code a lot simpler.

Hey there! Glad I could help!

I think it would be interesting to do a time execution comparison between both methods though, if you have the time.

Your other implementation was nice, complete and probably a bit faster, and SciPy values execution speed! 

But of course, if it turned out to be less fast than this, while being a bit more complicated, then I would replace it too.

**Comment by evbernardes** [comment]
> I could use a hand debugging the docs build error from someone more familiar with Sphinx. Even when I build locally verbose I don't get more than this, and for the life of me I cannot figure out where the error is.
> 
> ```
> /home/circleci/repo/doc/docstring of scipy.spatial.transform._rotation.Rotation:233:<autosummary>:1: WARNING: py:obj reference target not found: scipy.spatial.transform.Rotation.__pow__
> ```

I'm not an expert on Sphinx, so I can't help you with **why** this is happening.

But I noticed that a reference for `__pow__` isn't being created:

pow_not_link

So it is listed, but isn't a link.

File: scipy/spatial/transform/_rotation.pyx
Comment (evbernardes) [inline scipy/spatial/transform/_rotation.pyx:5]: You can remove this import

File: tools/lint.py
Comment (evbernardes) [inline tools/lint.py:85]: Is it normal that the linter was modified? Doesn't seem like something that should be in this PR

Comment (j-bowhay) [reply tools/lint.py:85]: This won't be needed once #17917 is merged

Comment (scottshambaugh) [reply tools/lint.py:85]: The above is merged, so this is no longer a diff.

**Comment by melissawm** [comment]
I found it - because this is a special method, you have to intentionally include it into our autosummary templates. In the file https://github.com/scipy/scipy/blob/main/doc/source/_templates/autosummary/class.rst you can edit the lines like so: 

Instead of
```
{% block methods %}
   .. HACK -- the point here is that we don't want this to appear in the output, but the autosummary should still generate the pages.
      .. autosummary::
         :toctree:
      {% for item in all_methods %}
         {%- if not item.startswith('_') or item in ['__call__', '__mul__', '__getitem__', '__len__'] %}
         {{ name }}.{{ item }}
         {%- endif -%}
      {%- endfor %}
      {% for item in inherited_members %}
         {%- if item in ['__call__', '__mul__', '__getitem__', '__len__'] %}
         {{ name }}.{{ item }}
         {%- endif -%}
      {%- endfor %}
  {% endblock %}
```

You want to have
```
{% block methods %}
   .. HACK -- the point here is that we don't want this to appear in the output, but the autosummary should still generate the pages.
      .. autosummary::
         :toctree:
      {% for item in all_methods %}
         {%- if not item.startswith('_') or item in ['__call__', '__mul__', '__getitem__', '__len__', '__pow__'] %}
         {{ name }}.{{ item }}
         {%- endif -%}
      {%- endfor %}
      {% for item in inherited_members %}
         {%- if item in ['__call__', '__mul__', '__getitem__', '__len__', '__pow__'] %}
         {{ name }}.{{ item }}
         {%- endif -%}
      {%- endfor %}
  {% endblock %}
```

**Comment by scottshambaugh** [comment]
Thank you @melissawm! I don’t think that’s something I would have found on my own, and it looks like that did the trick!

**Comment by scottshambaugh** [comment]
I've done a final pass over the implementation, tests, and docs, and think this is ready for merge.

**Comment by evbernardes** [comment]
In the generated docs, searching for `__pow__` returns 36 results and apart from the one for `Rotation.__pow__` they seem to be all empty, which I imagine is why `class.rst` was how it was in the first place.

For comparison, searching for `__pow__` in other versions returns no results, so maybe it should be checked if this is not desired.

**Comment by evbernardes** [comment]
Depending on what is the most desired solution for the docs, one solution could be to rename the current `__pow__` implementation to simply `pow`, which would have the docstring, and implement as new `__pow__` without any docstring as a simple call to `pow`. 

This would also solve the problem of having to listen `modulus` as a parameter.

**Comment by scottshambaugh** [comment]
Interesting. It seems like that is currently an issue for the other special methods such as `__len__`. Searching for that in the docs also turns up a bunch of results, but only about half of them are real methods. The good news is that these unused methods do not show up as methods under the doc pages for each object, so you'll only see this via search. I think renaming `__pow__` to `pow` and adding a wrapper function gets messy, and confusing for the docs since it's then not clear how the `**` operator is supported. 

IMO, this is an existing issue with the hacky bit of code melissawm pointed to above, and should be fixed separately rather than finding a workaround in this PR. I'll open an issue for it.

**Comment by scottshambaugh** [comment]
Deleted my previous comment as it was a red herring.

Interesting. I looked at a few of the examples in the search, and they actually *do* have the `__pow__` operator defined. To pick just one example: https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/_interface.py#L443-L447

So the issue is not that this is adding fake `__pow__` methods to the search results, but that existing methods were previously not showing up. I think that's a whole separate issue and I'll open a ticket for it.

**Comment by evbernardes** [comment]
> Interesting. I looked at a few of the examples in the search, and they actually *do* have the `__pow__` operator defined. To pick just one example: https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/_interface.py#L443-L447

Indeed! But these functions, even though they are defined, don't really have docstrings, which made me think not making pages for them was a design decision

**Comment by scottshambaugh** [comment]
Yeah, the pattern of what is/isn't documented there is not clear to me. I've opened up this ticket to follow up: https://github.com/scipy/scipy/issues/17953

**Comment by scottshambaugh** [comment]
@j-bowhay ok to add this to the 1.11.0 milestone?

**Comment by evbernardes** [comment]
One possible solution would be to remove both the changes to the templates and the docstring explaining `__pow__`, keeping the PR to a minimum so it's easier to get merged. 

Then, open another PR changing the template and adding docstrings to all `__pow__` (or other) implementations, since this is a different/possibly bigger issue.

What do you think @scottshambaugh?

**Comment by scottshambaugh** [comment]
This PR as currently written matches the conventions existing the library. I don't see resolution of that other ticket as a blocker to this PR, and don't really think we should intermix the two. Open to doing so if a maintainer thinks otherwise, but would like to keep as-is for now.

**Comment by evbernardes** [comment]
> Open to rolling in changes here now if a maintainer thinks otherwise, but would like to keep as-is currently.

Fair enough

As a user though (not as a curious developer) I don't think I'd ever look for the implementation of `__pow__` to see if a class supports exponentiation: I'd probably just look at the main docstring for the class itself.

**Comment by tylerjereddy** [comment]
Branching for the next release is imminent, and at this point I'll have to lean on core developers literally approving and merging themselves, because my bandwidth is limited and the release schedule will slip a lot if it runs past next week for branching/RC1.

So, I'm going to bump the milestone. I would also note that you may want to touch base with @nmayorov on some of your rotation PRs if you haven't already--IIRC they helped mentor the project that added the rotation machinery in the first place.

**Comment by scottshambaugh** [comment]
Thanks @tylerjereddy. If he's not available, I'm fairly lost on where/how/who to request reviews on this from. I've shot [two messages out to the mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/thread/CYQFDKGK7LGEN2E7TDD3ZP4ONRTE5PIV/#CYQFDKGK7LGEN2E7TDD3ZP4ONRTE5PIV) and these have been sitting open for a little over half a year, so I'm unsure what levers there are to push next. Any advice?

**Comment by scottshambaugh** [comment]
Hi @nmayorov, I have this and two other MR's in the scipy rotation module that are ready for review: https://github.com/scipy/scipy/pulls?q=is%3Aopen+is%3Apr+author%3Ascottshambaugh+rotation

I understand that you were part of the initial development there and are familiar with this section of the library, would you be able to take a look over these and share your thoughts? Thank you!

**Review by nmayorov — CHANGES_REQUESTED** [review]
Sorry for a late reply, let's finish it.

I think it's well done overall. I've commented on some minor things.

In general, why do you think it's useful to have a power function for `Rotation`? Just to know your reasoning.

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx]: updated

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx]: trimmed this down

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:2241]: My main thought here was to get ahead of any floating point errors that stack up in `as_rotvec` and `from_rotvec`, at least for the simplest cases where we can easily return an exact answer. I expect it might be marginally faster too.

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx]: Moved this up and made the negative `n` vs `-n` distinction unambiguous.

File: scipy/spatial/transform/_rotation.pyx
Comment (scottshambaugh) [reply scipy/spatial/transform/_rotation.pyx:1091]: I could open a PR just for this and the other Taylor series comments... seems such a small change though. The equations they are commenting on are very much inscrutable if you don't know beforehand that they are Taylor series expansions so I think the comments are necessary to put in at some point at least.

**Comment by scottshambaugh** [comment]
Thank you @nmayorov for the review! 

My primary motivation in adding powers here is that in my own work I am often trying to break up rotations into smaller pieces, i.e. finding the orientation that represents 1/3 of a full rotation. ``q = p ** (1/3)`` is a clean shorthand for that IMO - much tidier than ``q = Rotation.from_rotvec(1/3 * p.as_rotvec())`` and does not require the user to have an understanding of the rotation vector representation (which they might be unfamiliar with). The ``__pow__`` operator is also something that I find a very natural extension of the ``__mul__`` operator already in use here, and behaves as I would expect. I tried to use powers when first using the rotation module here, and was surprised that the behavior was not implemented - hence the impetus to add it :)

If you have time, I would also much appreciate if you could take a look at my other two open PRs in the rotation module:
* https://github.com/scipy/scipy/pull/17460
* https://github.com/scipy/scipy/pull/17542

File: scipy/spatial/transform/_rotation.pyx
Comment (nmayorov) [reply scipy/spatial/transform/_rotation.pyx:1091]: It's fine to keep it.

File: scipy/spatial/transform/_rotation.pyx
Comment (nmayorov) [reply scipy/spatial/transform/_rotation.pyx:2240]: Good point as well.

**Review by nmayorov — APPROVED** [review]

**Comment by nmayorov** [comment]
Looks good, documentation in HTML has no issues too (except for missing arguments for special methods from pyx/pyi, but it's unrelated).  I'm merging then.

I'll check your 2 other pull requests shortly.

**Pull request merged**
